pkg/didcomm/packager/packager.go

```diff
@@ -361,58 +374,77 @@ func (bp *Packager) getCTYAndPacker(envelope *transport.Envelope) (string, packe
+	for j, ka := range docResolution.DIDDocument.KeyAgreement {
 		kaID := ka.VerificationMethod.ID[strings.Index(ka.VerificationMethod.ID, "#")+1:]
-		if strings.EqualFold(kaID, keyAgrID[i+1:]) {
-			var (
-				marshalledKey []byte
-				recKey        *crypto.PublicKey
-			)
-
-			switch ka.VerificationMethod.Type {
-			case jsonWebKey2020:
-				jwkKey := ka.VerificationMethod.JSONWebKey()
+		if strings.EqualFold(kaID, keyAgrFragment) {
+			return marshalKeyFromVerificationMethod(keyAgrID, &ka.VerificationMethod, j)
+		}
 
-				recKey, err = jwksupport.PublicKeyFromJWK(jwkKey)
-				if err != nil {
-					return nil, fmt.Errorf("resolveKeyAgreementFromDIDDoc: for recipient JWK to PubKey %d: %w", i+1, err)
-				}
+		logger.Debugf("skipping keyID %s since it's not found in didDoc.KeyAgreement of did %s", kaID, keyAgrDID)
+	}
 
-				// for packing purposes, recipient kid header is the keyAgreement.ID value, the packers must resolve it and extract
-				// the kms kid value during unpack.
-				recKey.KID = keyAgrID
-			case x25519KeyAgreementKey2019:
-				recKey = &crypto.PublicKey{
-					KID:   keyAgrID,
-					X:     ka.VerificationMethod.Value,
-					Curve: "X25519",
-					Type:  "OKP",
-				}
-			default:
-				return nil, fmt.Errorf("resolveKeyAgreementFromDIDDoc: invalid KeyAgreement type %d: %T", i+1,
-					ka.VerificationMethod.Type)
-			}
+	for j := range docResolution.DIDDocument.VerificationMethod {
+		vm := &docResolution.DIDDocument.VerificationMethod[j]
 
-			marshalledKey, err = json.Marshal(recKey)
-			if err != nil {
-				return nil, fmt.Errorf("resolveKeyAgreementFromDIDDoc: for recipient DID keyAgreement key marshal %d: %w",
-					i+1, err)
-			}
+		vmID := vm.ID[strings.Index(vm.ID, "#")+1:]
+		logger.Infof("vm: %#v", vm)
 
-			return marshalledKey, nil
+		if strings.EqualFold(vmID, keyAgrFragment) {
+			return marshalKeyFromVerificationMethod(keyAgrID, vm, j)
 		}
-
-		logger.Debugf("skipping keyID %s since it's not found in didDoc.KeyAgreement of did %s", kaID, keyAgrID[:i])
 	}
 
 	return nil, fmt.Errorf("resolveKeyAgreementFromDIDDoc: keyAgreement ID '%s' not found in DID '%s'", keyAgrID,
 		docResolution.DIDDocument.ID)
 }
+
+func marshalKeyFromVerificationMethod(keyAgrID string, vm *did.VerificationMethod, i int) (*crypto.PublicKey, error) {
+	var (
+		recKey *crypto.PublicKey
+		err    error
+	)
+
+	switch vm.Type {
+	case jsonWebKey2020:
+		jwkKey := vm.JSONWebKey()
+
+		recKey, err = jwksupport.PublicKeyFromJWK(jwkKey)
+		if err != nil {
+			return nil, fmt.Errorf("resolveKeyAgreementFromDIDDoc: for recipient JWK to PubKey %d: %w", i+1, err)
+		}
+
+		// for packing purposes, recipient kid header is the keyAgreement.ID value, the packers must resolve it and extract
+		// the kms kid value during unpack.
+		recKey.KID = keyAgrID
+	case x25519KeyAgreementKey2019:
+		recKey = &crypto.PublicKey{
+			KID:   keyAgrID,
+			X:     vm.Value,
+			Curve: "X25519",
+			Type:  "OKP",
+		}
+	case "Ed25519VerificationKey2018":
```

Comment:
This condition is for legacy packer only which is based on recipientKey as base58 encoded keys coming from destination.RecipientKeys, the packager must not get DID doc's VMs for legacy packing. Let's not propagate legacy stuff further in the framework.